src/matchers/text.rs

```diff
@@ -0,0 +1,91 @@
+/// Returns whether a buffer is html data.
+///
+/// Conforms to [whatwg](https://mimesniff.spec.whatwg.org/)
+/// specification.
+pub fn is_html(buf: &[u8]) -> bool {
+    let values: &[&[u8]] = &[
+        b"<!DOCTYPE HTML",
+        b"<HTML",
+        b"<HEAD",
+        b"<SCRIPT",
+        b"<IFRAME",
+        b"<H1",
+        b"<DIV",
+        b"<FONT",
+        b"<TABLE",
+        b"<A",
+        b"<STYLE",
+        b"<TITLE",
+        b"<B",
+        b"<BODY",
+        b"<BR",
+        b"<P",
+        b"<!--",
+    ];
+    let buf = trim_start_whitespaces(buf);
+
+    for val in values {
+        if starts_with_ignore_ascii_case(buf, val) && buf.len() > val.len() {
+            match buf[val.len()] {
+                // tag-terminitating byte
+                0x20 | 0x3E => return true,
+                _ => continue,
+            }
+        }
+    }
+
+    false
+}
+
+/// Returns whether a buffer is xml data.
+///
+/// Conforms to [whatwg](https://mimesniff.spec.whatwg.org/)
+/// specification.
+pub fn is_xml(buf: &[u8]) -> bool {
+    let val: &[u8] = b"<?xml";
+    let buf = trim_start_whitespaces(buf);
+    starts_with_ignore_ascii_case(buf, val)
+}
+
+/// Strip whitespaces at the beginning of the buffer.
+///
+/// Follows https://mimesniff.spec.whatwg.org
+/// definition of whitespace.
+fn trim_start_whitespaces(buf: &[u8]) -> &[u8] {
+    for (i, b) in buf.iter().enumerate() {
+        match b {
+            0x09 | 0x0A | 0x0C | 0x0D | 0x20 => continue,
+            _ => return &buf[i..],
+        }
+    }
+    &[]
+}
+
+fn starts_with_ignore_ascii_case(buf: &[u8], needle: &[u8]) -> bool {
```

Comment:
Placing #[inline] over `trim_start_whitespaces` and `trim_start_whitespaces` seems to make it generate a bit less code, haven't tested the performance

Reply:
I would have expected the compiler to do the job himself or have good reasons not to do it. If we refer to [inline doc](https://doc.rust-lang.org/reference/attributes/codegen.html#the-inline-attribute) I believe we should not introduce compiler hint unless justified by bench results. I may be wrong but I would be surprised that this optimization brings any value as far as binary size or performances are concerned. I am a kind of "early optimization is route of all evil" guy ...

Reply:
I tested this again and it's already inlining them